src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -19,152 +20,274 @@
 #include "storage/table.hpp"
 #include "utils/assert.hpp"
 
-namespace opossum {
-
-void ChunkPruningRule::_apply_to_plan_without_subqueries(const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
-  // we only want to follow chains of predicates
-  if (lqp_root->type != LQPNodeType::Predicate) {
-    _apply_to_plan_inputs_without_subqueries(lqp_root);
-    return;
-  }
+namespace {
+using namespace opossum;  // NOLINT
+using PredicatePruningChain = std::vector<std::shared_ptr<PredicateNode>>;
+
+/**
+ * This function traverses the LQP upwards from @param next_node to find all predicates that filter
+ * @param stored_table_node. If the LQP branches, recursion is used to continue @param current_predicate_pruning_chain
+ * for all branches.
+ *
+ * @returns all predicate pruning chain(s) that filter @param stored_table_node. Note, that these chains can overlap
+ * in their first few nodes.
+ */
+std::vector<PredicatePruningChain> find_predicate_pruning_chains_by_stored_table_node_recursively(
+    const std::shared_ptr<AbstractLQPNode>& next_node, PredicatePruningChain current_predicate_pruning_chain,
+    const std::shared_ptr<StoredTableNode>& stored_table_node,
+    std::unordered_set<std::shared_ptr<AbstractLQPNode>> visited_nodes) {
```

Comment:
If we pass in visited_nodes by copy, won't the left and the right path of a diamond have two independent versions? This almost looks as if the corresponding assertion can never fail.

Reply:
True! I changed it to pass by reference.